iree/compiler/Dialect/Flow/IR/FlowOps.cpp

```diff
@@ -263,6 +265,118 @@ static LogicalResult verifyVariableStoreIndirectOp(
 // flow.dispatch.region
 //===----------------------------------------------------------------------===//
 
+llvm::Optional<std::pair<DispatchRegionOp, Operation *>>
+DispatchRegionOp::formFromAnchorOp(Value workload, Operation *anchorOp,
+                                   OpBuilder &builder) {
+  builder.setInsertionPoint(anchorOp);
+  auto loc = anchorOp->getLoc();
+  // Map anchor into new dispatch region.
+  llvm::SmallVector<Value, 4> capturedInputs(anchorOp->getOperands());
+  llvm::SmallVector<Type, 1> types(anchorOp->getResultTypes().begin(),
+                                   anchorOp->getResultTypes().end());
+  auto drOp =
+      builder.create<DispatchRegionOp>(loc, types, workload, capturedInputs);
+  auto *drBlock = new Block();
+  drOp.body().push_back(drBlock);
+  BlockAndValueMapping mapping;
+  for (Value capturedInput : capturedInputs) {
+    auto blockArg = drBlock->addArgument(capturedInput.getType());
+    mapping.map(capturedInput, blockArg);
+  }
+
+  // Create new body.
+  builder.setInsertionPointToEnd(drBlock);
+  auto *newAnchorOp = builder.clone(*anchorOp, mapping);
+  builder.create<IREE::Flow::ReturnOp>(loc, newAnchorOp->getResults());
+
+  // Replace anchor uses with region result.
+  for (auto it : llvm::enumerate(anchorOp->getResults())) {
+    it.value().replaceAllUsesWith(drOp.getResult(it.index()));
+  }
+  anchorOp->erase();
+  return std::make_pair(drOp, newAnchorOp);
+}
+
+void DispatchRegionOp::dceOperandsAndResults(DispatchRegionOp &op) {
+  OpBuilder builder(op.getContext());
```

Comment:
ah here's your non-PatternRewriter use; still would be nice to have a single static method for this (optimizeClosureOpInPlace or something)